internal/k8s/helpers.go

```diff
@@ -83,3 +89,77 @@ func isStatusEqual(objA, objB interface{}) bool {
 	}
 	return false
 }
+
+// IsObjectEqual checks if objects received during update are equal.
+//
+// Make an attempt to avoid comparing full objects since it can be very CPU intensive.
+// Prefer comparing Generation when only interested in spec changes.
+func IsObjectEqual(old, new interface{}) (bool, error) {
+
+	// Fast path for any object: when ResourceVersions are equal, the objects are equal.
+	// NOTE: This optimizes the case when controller-runtime executes full sync and sends updates for all objects.
+	if isResourceVersionEqual(old, new) {
+		return true, nil
+	}
+
+	switch old := old.(type) {
+
+	// Fast path for objects that implement Generation and where only spec changes matter.
+	// Generation is implemented in CRDs, Ingress and IngressClass.
+	case *contour_api_v1alpha1.ExtensionService:
+		return isGenerationEqual(old, new), nil
+	case *contour_api_v1.TLSCertificateDelegation:
+		return isGenerationEqual(old, new), nil
+	case *contour_api_v1alpha1.ContourConfiguration:
+		return isGenerationEqual(old, new), nil
+
+	case *gatewayapi_v1beta1.GatewayClass:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1beta1.Gateway:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1beta1.HTTPRoute:
+		return isGenerationEqual(old, new), nil
+	case *gatewayapi_v1alpha2.TLSRoute:
+		return isGenerationEqual(old, new), nil
+
+	// Slow path: compare the content of the objects.
+	case *contour_api_v1.HTTPProxy:
+		if new, ok := new.(*contour_api_v1.HTTPProxy); ok {
+			return isGenerationEqual(old, new) &&
+				apiequality.Semantic.DeepEqual(old.ObjectMeta.Annotations, new.ObjectMeta.Annotations), nil
+		}
+	case *networking_v1.Ingress:
+		if new, ok := new.(*networking_v1.Ingress); ok {
+			return isGenerationEqual(old, new) &&
+				apiequality.Semantic.DeepEqual(old.ObjectMeta.Annotations, new.ObjectMeta.Annotations), nil
+		}
+	case *v1.Secret:
+		if new, ok := new.(*v1.Secret); ok {
+			return reflect.DeepEqual(old.Data, new.Data), nil
+		}
+	case *v1.Service:
+		if new, ok := new.(*v1.Service); ok {
+			return apiequality.Semantic.DeepEqual(old.Spec, new.Spec) &&
+				apiequality.Semantic.DeepEqual(old.Status, new.Status) &&
+				apiequality.Semantic.DeepEqual(old.ObjectMeta.Annotations, new.ObjectMeta.Annotations), nil
+		}
+	case *v1.Endpoints:
+		if new, ok := new.(*v1.Endpoints); ok {
+			return apiequality.Semantic.DeepEqual(old.Subsets, new.Subsets), nil
+		}
+
+	default:
+
+	}
+
+	// ResourseVersions are not equal and we don't know how to compare the object type.
+	return false, fmt.Errorf("do not know how to compare %s", old.(client.Object).GetObjectKind().GroupVersionKind().String())
```

Comment:
Do you think it's necessary to return an error here? We're effectively falling back to ResourceVersion comparison, which seems fine. I guess the question is, does logging this have any benefit for the user/operator?

Reply:
I was thinking how to alert the developer who might miss adding detailed comparison when introducing new type. I wonder, can it be said that for every handled type, there should be explicit comparison? 

If yes, then without error printout, it might be likely that we only notice the problem much later e.g. when performance impact becomes inevitable.

If no, then I can remove the error.

Reply:
> I was thinking how to alert the developer who might miss adding detailed comparison when introducing new type. I wonder, can it be said that for every handled type, there should be explicit comparison?

I think this is probably right. In that case, I wonder if it'd be better to use a panic on an unknown object type? This would cause E2E tests to fail, which would then force the code to be updated before getting merged, whereas the error log could always be missed.

As an illustrative example, I know we need to add Namespace and GRPCRoute to this function still, but they are only used for Gateway API so you wouldn't even see the errors in the log unless you were testing with Gateway API. But, with a panic in place, the E2E's would fail.

I know intentionally adding `panic()` to code doesn't sit right with everyone, so curious what folks think.

Another option would be to look to enforce this at compile time, by introducing an interface that all types we cache would be required to implement. I'd like to look into this some more to see if/how it could work, but it doesn't need to be done as part of this PR.

Reply:
> Another option would be to look to enforce this at compile time, by introducing an interface that all types we cache would be required to implement. I'd like to look into this some more to see if/how it could work, but it doesn't need to be done as part of this PR.

yeah this is sorta what i was heading towards with the generics approach

we would probably need to add some wrapper types of our own around the k8s core resources or GW API resources if we need to get at some of the internals (spec field, status field, etc.) with a `.Equal` method or something similar and make that interface a constraint

the way i did it doesn't go this far ^ and still has switches on types in the comparison helper, so could be done better with some more thinking

Reply:
@sunjayBhatia yeah I like that idea. I'm fine with the impl as-is for now as long as we're looking into some improvements here down the road, since I do think it's super-easy to miss all the various spots to update for new types right now. But, in this case, falling back to ResourceVersion comparison isn't so bad either.

Reply:
Thank You @sunjayBhatia and @skriss for looking into this!

I must confess that I have not looked much into Go generics yet, so I'm likely not knowledgeable enough to grasp how it could be done :) To me it seems impossible to influence the interfaces in such a way that the problem is converted into compiler time error, because the objects are Kubernetes types, constructed by controller-runtime.

Reply:
yeah for sure, the controller-runtime/client-go controller/reconciler/handler interfaces all use `interface{}` there will likely be at the very least 1 place we need to do a type assertion